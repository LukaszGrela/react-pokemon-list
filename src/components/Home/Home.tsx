import React, { useState } from 'react';
import { useParams } from 'react-router-dom';
import Modal from '../Modal/Modal';
import PokemonList from '../PokemonList/PokemonList';
import './style/index.scss';

const Home: React.FC = (): JSX.Element => {
  const { page } = useParams();
  const [showModal, setShowModal] = useState(false)

  const handlePokemonSelection = (pid: string): void => {
    console.log('', pid);
    setShowModal(true)
  }

  return (
    <main className='Home'>
      <header>
        <h1>Pok√©mon List</h1>
      </header>
      <PokemonList page={page ? parseInt(page) : 1} interactive={!showModal} handlePokemonSelect={handlePokemonSelection} />
      <footer>{page}</footer>
      <Modal
        isOpen={showModal}
        closeModal={() => {
          setShowModal(false)
        }}
      >Modal</Modal>
    </main>
  );
};

export default Home;
