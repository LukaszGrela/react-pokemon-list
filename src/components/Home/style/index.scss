@import 'styles/exports';

@import 'components/Home/style/Home';
