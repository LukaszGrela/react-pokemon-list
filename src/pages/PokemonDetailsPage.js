import React from 'react';
import { connect } from 'react-redux';

import PokemonDetails from '../components/PokemonDetails';

import './styles/PokemonDetailsPage.scss';
import { actionGetPokemonDetails } from '../actions/actionPokemonDetails';
import { pokemonListSelector } from '../selectors/pokemonListSelector';

class PokemonDetailsPage extends React.Component {

    render = () => {
        const { id, name, history, pullPokemonDetails } = this.props;
        return (
            <article className='pokemon-details-page'>
                <header>{name}</header>
                <PokemonDetails
                    id={parseInt(id, 10)}
                    history={history}
                    handleShowPokemon={pullPokemonDetails}
                />
            </article>
        );
    }
};

const mapStateToProps = (state, props) => {
    const { match } = props;
    const id = parseInt(match.params.id, 10);
    const pokemon = pokemonListSelector(state.pokemons.list, id);
    const { name } = pokemon;
    return {
        id,
        name
        // loading: state.pokemons.loading,
        // error: state.pokemons.error,
        // noMore: state.pokemons.noMore,
        // list: state.pokemons.list
    }
};
const mapDispatchTopProps = (dispatch, { history }) => ({
    /* action to pull pokemon data  */
    pullPokemonDetails: (id) => {
        history.push(`/pokemon/${id}`);
        return dispatch(actionGetPokemonDetails(id));
    }
})
export default connect(mapStateToProps, mapDispatchTopProps)(PokemonDetailsPage);