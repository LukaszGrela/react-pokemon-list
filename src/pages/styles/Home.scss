@import '../../styles/_mixins.scss';

.home-page {
    text-align: center;
    @include media(0, 41rem) {
        
            width: 100%;
        
    }
    .loading-spinner {
        width: 100%;
        margin-bottom: 3rem;
    }
    .no-more {
        display: block;
        width: 100%;
        font-size: 1.6rem;
        margin: 2rem 0;
        margin-bottom: 3rem;

        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
        p {
            padding: 0 0.8rem;
        }
    }
}