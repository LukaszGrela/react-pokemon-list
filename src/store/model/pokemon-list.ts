import { TNamedAPIResource } from "./common";

export interface IAPIResourceList {
  count: number;
  next: string;
  previous: string | null;
  results: TNamedAPIResource[];
}
