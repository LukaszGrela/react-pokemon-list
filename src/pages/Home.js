import React from 'react';
import { connect } from 'react-redux';
import { Switch, Route } from 'react-router-dom';
import { actionGetPokemonList } from '../actions/actionPokemonList';
import PokemonList from '../components/PokemonList';
import Spinner from '../components/Spinner';

import './styles/Home.scss';
import Panel from '../components/Panel';
import PokemonDetails from '../components/PokemonDetails';

export class Home extends React.Component {

    previousLocation = this.props.location;

    handleListItemClick = (id) => {
        this.props.history.push(`/pokemon/${id}`, { modal: true });
        
    }

    handleScroll = () => {
        if ((window.innerHeight + window.scrollY) >= document.body.scrollHeight) {
            // reached the bottom
            this.detachScrollListener();//stop listening
            this.props.pullMoreItems();
        }
    }

    detachScrollListener = () => {
        window.removeEventListener('scroll', this.handleScroll);
    }
    attachScrollListener = () => {
        this.detachScrollListener();
        window.addEventListener('scroll', this.handleScroll);
    }

    componentWillUpdate = (nextProps) => {
        const { location } = this.props;
        // set previousLocation if props.location is not modal
        if (
            nextProps.history.action !== "POP" &&
            (!location.state || !location.state.modal)
        ) {
            this.previousLocation = this.props.location;
        }
    }
    componentDidUpdate = (prevProps, prevState) => {
        if (this.props.list.length > 0 //dont for initial load
            && this.props.loading === true //is loading
            && this.props.loading !== prevProps.loading //was just switched to loading
        ) {
            const spinner = document.getElementsByClassName('loading-spinner');
            if(spinner && spinner.length > 0) spinner[0].scrollIntoView();
        }
        if (this.props.noMore !== prevProps.noMore && this.props.noMore) {
            this.detachScrollListener();//no more pulling
        }
        if (!this.props.noMore && this.props.loading !== prevProps.loading && this.props.loading === false) {
            this.attachScrollListener();//reattach listener
        }
    }
    componentWillMount = () => {
        this.attachScrollListener();
    }
    componentWillUnmount = () => {
        this.detachScrollListener();
    }
    render = () => {
        const { location, list, loading, noMore } = this.props;
        const isModal = !!(
            location.state &&
            location.state.modal &&
            this.previousLocation !== location
        ); // not initial render
        return (
            <article className='home-page'>

                <Switch location={isModal ? this.previousLocation : location}>
                    <Route exact path='/' component={
                        () => (
                            <PokemonList list={list} handleClick={this.handleListItemClick} />
                        )
                    } />
                </Switch>
                {loading && <Spinner />}
                {noMore &&
                    <div className='no-more'>
                        <p>-- The End --</p>
                    </div>
                }
                <Route exact path='/pokemon/:id' component={
                    ({ history, match }, b) => {
                        return (
                            <Panel handlePanelDismiss={ _=>{
                                history.replace('/');
                            }} >
                                <PokemonDetails id={parseInt(match.params.id,10)} />
                            </Panel>
                        );
                    }
                } />
            </article>
        );
    }
};

const mapStateToProps = (state, props) => ({
    loading: state.pokemons.loading,
    error: state.pokemons.error,
    noMore: state.pokemons.noMore,
    list: state.pokemons.list
});
const mapDispatchTopProps = (dispatch) => ({
    pullMoreItems: () => dispatch(actionGetPokemonList())
})
export default connect(mapStateToProps, mapDispatchTopProps)(Home);