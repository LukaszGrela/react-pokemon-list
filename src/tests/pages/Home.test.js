import React from 'react';
import { shallow } from 'enzyme';
import { Home } from '../../pages/Home';
import { POKEMON_LIST_RESULTS } from '../fixtures/pokemon-list';

let wrapper,
    location,
    history,
    loading,
    error,
    noMore,
    list,
    pullMoreItems;

beforeEach(() => {
    history = { action: 'PUSH' };
    location = { state: { modal: false } };
    loading = false;
    error = null;
    noMore = false;
    list = POKEMON_LIST_RESULTS;
    pullMoreItems = jest.fn();
    wrapper = shallow(<Home
        location={location}
        history={history}
        loading={loading}
        error={error}
        noMore={noMore}
        list={list}
        pullMoreItems={pullMoreItems}
    />);
})

test('Should render Home page correctly', () => {

    expect(wrapper).toMatchSnapshot();
});
test('Should render Home page correctly for loading state', () => {
    wrapper.setProps(
        {
            loading: true
        });

    expect(wrapper).toMatchSnapshot();
});
test('Should render Home page correctly for noMore state', () => {
    wrapper.setProps(
        {
            noMore: true
        });

    expect(wrapper).toMatchSnapshot();
});