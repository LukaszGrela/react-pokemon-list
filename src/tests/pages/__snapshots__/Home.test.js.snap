// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Home page correctly 1`] = `
<article
  className="home-page"
>
  <Switch
    location={
      Object {
        "state": Object {
          "modal": false,
        },
      }
    }
  >
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
  </Switch>
  <LoadMore
    clickHandler={[Function]}
  />
  <Route
    component={[Function]}
    exact={true}
    path="/pokemon/:id"
  />
</article>
`;

exports[`Should render Home page correctly for loading state 1`] = `
<article
  className="home-page"
>
  <Switch
    location={
      Object {
        "state": Object {
          "modal": false,
        },
      }
    }
  >
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
  </Switch>
  <Spinner />
  <Route
    component={[Function]}
    exact={true}
    path="/pokemon/:id"
  />
</article>
`;

exports[`Should render Home page correctly for noMore state 1`] = `
<article
  className="home-page"
>
  <Switch
    location={
      Object {
        "state": Object {
          "modal": false,
        },
      }
    }
  >
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
  </Switch>
  <LoadMore
    clickHandler={[Function]}
  />
  <div
    className="no-more"
  >
    <p>
      -- The End --
    </p>
  </div>
  <Route
    component={[Function]}
    exact={true}
    path="/pokemon/:id"
  />
</article>
`;
